src/terms.rs

```diff
@@ -0,0 +1,529 @@
+                let func_type = self
+                    .type_info()
+                    .lookup_expr(self.current_ctx, &NormExpr::Call(*head, body.clone()))
+                    .unwrap();
+                if func_type.is_datatype {
+                    let view_name = self.view_name(*head);
+                    NormFact::Assign(*lhs, NormExpr::Call(view_name, body.clone()))
+                } else {
+                    fact.clone()
+                }
+            }
+            _ => fact.clone(),
+        }
+        .map_use(&mut |var| {
+            let vtype = self.type_info().lookup(self.current_ctx, var).unwrap();
+            if !self.type_info().is_global(var) {
+                Expr::Var(var)
+            } else if let Some(wrapped) = self.wrap_parent_or_rebuild(var.to_string(), vtype) {
+                self.desugar().expr_parser.parse(&wrapped).unwrap()
+            } else {
+                Expr::Var(var)
+            }
+        })
+    }
+
+    fn wrap_parent_or_rebuild(&self, var: String, sort: ArcSort) -> Option<String> {
+        if sort.is_container_sort() {
+            Some(format!("(rebuild {})", var))
+        } else {
+            self.wrap_parent(var, sort)
+        }
+    }
+
+    fn wrap_parent(&self, var: String, sort: ArcSort) -> Option<String> {
+        // TODO make all containers also eq sort
+        if sort.is_eq_sort() {
+            let parent = self.parent_name(sort.name());
+            Some(format!("({parent} {var})"))
+        } else {
+            None
+        }
+    }
+
+    fn instrument_facts(&mut self, facts: &[NormFact]) -> Vec<Fact> {
+        facts.iter().map(|f| self.instrument_fact(f)).collect()
+    }
+
+    pub(crate) fn parse_actions(&self, actions: Vec<String>) -> Vec<Action> {
+        actions
+            .into_iter()
+            .map(|s| self.desugar().action_parser.parse(&s).unwrap())
+            .collect()
+    }
+
+    fn fresh_var(&mut self) -> Symbol {
+        self.egraph.desugar.get_fresh()
+    }
+
+    fn canonicalize_everything(
+        &mut self,
+        action: &NormAction,
+        res: &mut Vec<Action>,
+    ) -> NormAction {
+        action.map_def_use(&mut |var, is_def| {
+            let vtype = self.type_info().lookup(self.current_ctx, var).unwrap();
+
+            if is_def {
+                var
+            } else if let Some(wrapped) = self.wrap_parent_or_rebuild(var.to_string(), vtype) {
+                let fresh_var = self.fresh_var();
+                res.extend(
+                    self.desugar()
+                        .action_parser
+                        .parse(&format!("(let {fresh_var} {wrapped})")),
+                );
+                fresh_var
+            } else {
+                var
+            }
+        })
+    }
+
+    // Actions need to be instrumented to add to the view
+    // as well as to the terms tables.
+    // In addition, we need to look up canonical versions
+    // of globals.
```

Comment:
I don't see how globals are relevant to this function?

Reply:
In the terms encoding, I didn't want to rely on rebuilding. That means globals don't get canonicalized- they stay equal to a particular term.
To get around this, I make every rule look up the global every time.

However, this might make containers very slow! Containers must be rebuilt from the global every time the query is run.

Reply:
Thanks for the clarification